src/full/Agda/TypeChecking/Monad/Signature.hs

```diff
@@ -307,28 +310,27 @@ applySection new ptel old ts ScopeCopyInfo{ renModules = rm, renNames = rd } = d
     -- and if a constructor is copied its datatype needs to be.
     closeConstructors :: Ren QName -> TCM (Ren QName)
     closeConstructors rd = do
-        ds <- nubOn id . catMaybes <$> mapM (constructorData  . fst) rd
-        cs <- nubOn id . concat    <$> mapM (dataConstructors . fst) rd
-        new <- concat <$> mapM rename (ds ++ cs)
+        ds <- nubOn id . catMaybes <$> traverse constructorData (Map.keys rd)
+        cs <- nubOn id . concat    <$> traverse dataConstructors (Map.keys rd)
+        new <- Map.unionsWith (S.<>) <$> traverse rename (ds ++ cs)
         reportSDoc "tc.mod.apply.complete" 30 $
           "also copying: " <+> pretty new
-        return $ new ++ rd
+        return $ Map.unionWith (S.<>) new rd
       where
         rename :: QName -> TCM (Ren QName)
-        rename x =
-          case lookup x rd of
-            Nothing -> do y <- freshName_ (show $ qnameName x)
-                          return [(x, qnameFromList $ singleton y)]
-            Just{}  -> return []
+        rename x
+          | x `Map.member` rd = pure mempty
+          | otherwise =
+              Map.singleton x . pure . qnameFromList . singleton <$> freshName_ (show $ qnameName x)
 
         constructorData :: QName -> TCM (Maybe QName)
-        constructorData x = do
+        constructorData x =
```

Comment:
There is no need to remove these redundant `do`s.  I prefer to have them for uniformity.